go/vt/vtgate/planbuilder/operators/union_merging.go

```diff
@@ -180,25 +180,27 @@ func createMergedUnion(
 	cols := make(sqlparser.SelectExprs, len(lhsExprs))
 	noDeps := len(lhsExprs) != len(rhsExprs)
 	for idx, col := range lhsExprs {
-		ae, ok := col.(*sqlparser.AliasedExpr)
+		lae, ok := col.(*sqlparser.AliasedExpr)
 		if !ok {
 			cols[idx] = col
 			noDeps = true
 			continue
 		}
-		col := sqlparser.NewColName(ae.ColumnName())
+		col := sqlparser.NewColName(lae.ColumnName())
 		cols[idx] = aeWrap(col)
 		if noDeps {
 			continue
 		}
 
-		deps := ctx.SemTable.RecursiveDeps(ae.Expr)
-		ae, ok = rhsExprs[idx].(*sqlparser.AliasedExpr)
+		deps := ctx.SemTable.RecursiveDeps(lae.Expr)
+		rae, ok := rhsExprs[idx].(*sqlparser.AliasedExpr)
 		if !ok {
 			noDeps = true
 			continue
 		}
-		deps = deps.Merge(ctx.SemTable.RecursiveDeps(ae.Expr))
+		deps = deps.Merge(ctx.SemTable.RecursiveDeps(rae.Expr))
+		ctx.SemTable.CopySemanticInfo(rae.Expr, col)
+		ctx.SemTable.CopySemanticInfo(lae.Expr, col)
```

Comment:
I don't think this is correct. This works when both the left and right columns have the same type, but if they have different types, we should coerce the resulting type and not just copy it.

Here is an example query:

```sql
mysql> select 1 union select 'a' union select now();
Field   1:  `1`
Catalog:    `def`
Database:   ``
Table:      ``
Org_table:  ``
Type:       VAR_STRING
Collation:  utf8mb4_0900_ai_ci (255)
Length:     76
Max_length: 19
Decimals:   0
Flags:      NOT_NULL 


+---------------------+
| 1                   |
+---------------------+
| 1                   |
| a                   |
| 2024-01-30 07:32:57 |
+---------------------+
3 rows in set (0.00 sec)
```

As you can see when you run the mysql client with the option `--column-type-info`, we have three columns with different types - int, varchar and datetime, and they are all coerced into varchar.

Fortunately, we already have this functionality today. Here is my suggestion:

```suggestion
		rt, foundR := ctx.SemTable.TypeForExpr(rae.Expr)
		lt, foundL := ctx.SemTable.TypeForExpr(lae.Expr)
		if foundR && foundL {
			types := []sqltypes.Type{rt.Type(), lt.Type()}
			t := evalengine.AggregateTypes(types)
			ctx.SemTable.ExprTypes[col] = evalengine.NewType(t, collations.Unknown)
		}
```

Reply:
Yes, I overlooked the case where the left and right columns have different types. However, in some cases, you may not be able to retrieve the ``evalengine.Type`` through ``ctx.SemTable.TypeForExpr()`` method, especially when dealing with syntax tree nodes like ``now()`` that cannot provide the evalengine.Type after processing it with ``ctx.SemTable.TypeForExpr()`` .